src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetDelegateForFunctionPointerTests.cs

```diff
@@ -0,0 +1,162 @@
+            methodBuilder.SetImplementationFlags(MethodImplAttributes.Runtime | MethodImplAttributes.Managed);
+
+            Type type = typeBuilder.CreateType();
+
+            Delegate functionDelegate = Marshal.GetDelegateForFunctionPointer(ptr, type);
+            Assert.NotNull(functionDelegate);
+            Assert.IsType<NonGenericDelegate>(functionDelegate);
+            Assert.Equal(targetMethod, functionDelegate.Method);
+            Assert.Null(functionDelegate.Target);
+        }
+
+        [Fact]
+        public void GetDelegateForFunctionPointer_Generic_ReturnsExpected()
+        {
+            MethodInfo targetMethod = typeof(GetDelegateForFunctionPointerTests).GetMethod(nameof(Method));
+            Delegate d = targetMethod.CreateDelegate(typeof(NonGenericDelegate));
+            IntPtr ptr = Marshal.GetFunctionPointerForDelegate(d);
+
+            Delegate functionDelegate = Marshal.GetDelegateForFunctionPointer(ptr, typeof(NonGenericDelegate));
+            Assert.NotNull(functionDelegate);
+            Assert.IsType<NonGenericDelegate>(functionDelegate);
+            Assert.Equal(targetMethod, functionDelegate.Method);
+            Assert.Null(functionDelegate.Target);
+        }
+
+        [Fact]
+        public void GetDelegateForFunctionPointer_GenericInvalidType_ReturnsExpected()
+        {
+            MethodInfo targetMethod = typeof(GetDelegateForFunctionPointerTests).GetMethod(nameof(Method));
+            Delegate d = targetMethod.CreateDelegate(typeof(NonGenericDelegate));
+            IntPtr ptr = Marshal.GetFunctionPointerForDelegate(d);
+
+            Delegate functionDelegate = Assert.IsType<NonGenericDelegate>(Marshal.GetDelegateForFunctionPointer<MulticastDelegate>(ptr));
+            Assert.NotNull(functionDelegate);
+            Assert.IsType<NonGenericDelegate>(functionDelegate);
+            Assert.Equal(targetMethod, functionDelegate.Method);
+            Assert.Null(functionDelegate.Target);
+        }
+
+        [Fact]
+        public void GetDelegateForFunctionPointer_ZeroPointer_ThrowsArgumentNullException()
+        {
+            AssertExtensions.Throws<ArgumentNullException>("ptr", () => Marshal.GetDelegateForFunctionPointer(IntPtr.Zero, typeof(NonGenericDelegate)));
+            AssertExtensions.Throws<ArgumentNullException>("ptr", () => Marshal.GetDelegateForFunctionPointer<string>(IntPtr.Zero));
+        }
+
+        [Fact]
+        public void GetDelegateForFunctionPointer_NullType_ThrowsArgumentNullException()
+        {
+            AssertExtensions.Throws<ArgumentNullException>("t", () => Marshal.GetDelegateForFunctionPointer((IntPtr)1, null));
+        }
+
+        public static IEnumerable<object[]> GetDelegateForFunctionPointer_InvalidType_TestData()
+        {
+            yield return new object[] { typeof(int).MakeByRefType() };
```

Comment:
can use TheoryData<U, V, W> feature of the Xunit. eg https://github.com/dotnet/corefx/blob/master/src/Common/tests/Tests/System/IO/PathInternal.Tests.cs#L68

Reply:
Can't do that, because it contains variable declarations:

```cs
            AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Assembly"), AssemblyBuilderAccess.Run);
            ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule("Module");
            TypeBuilder typeBuilder = moduleBuilder.DefineType("Type");
```

Reply:
you can declare variables inside it but you wont be able to inline it. Its like writing code inside the get property.
```C#
public static TheoryData<t> propname
{ 
     get
     { 
           var returnList = new TheoryData<t>;
           returnList.Add()
           // more code
          return returnList
     }
}
```

Reply:
I mean at that point I don't think it really is worth it - better to be consistent with the rest of the test code in InteropServices tests which doesn't use `TheoryData` and follows the pattern in this PR